### PR TITLE
fix: Fix completions disappearing when typing two keys in quick succession

### PR DESCRIPTION
**Pull request by Veykril**
With this PR we now retry requests if they get cancelled due to document changes.

This fixes the completions problem we have where completions seem to randomly disappear, see https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Completions.20not.20always.20appearing 

Fixes https://github.com/rust-lang/rust-analyzer/issues/10187
Fixes https://github.com/rust-lang/rust-analyzer/issues/7560
Fixes https://github.com/rust-lang/rust-analyzer/issues/12153

### CODE REVIEW

File: crates/rust-analyzer/src/dispatch.rs [hunk 1]
```diff
@@ -37,49 +37,55 @@ impl<'a> RequestDispatcher<'a> {
...
-            None => return Ok(self),
+            None => return self,
+        };
+        let result = {
+            let _pctx = stdx::panic_context::enter(panic_context);
+            f(self.global_state, params)
+        };
+        match result_to_response::<R>(req.id.clone(), result) {
+            Ok(response) => self.global_state.respond(response),
+            Err(_) => self.global_state.task_pool.handle.send_retry(req),
```
Comment (matklad) [inline crates/rust-analyzer/src/dispatch.rs:56]: This I think is unreachable: sync task can’t get cencellles

File: crates/rust-analyzer/src/main_loop.rs [hunk 2]
```diff
@@ -172,13 +173,13 @@ impl GlobalState {
...
             recv(self.task_pool.receiver) -> task =>
-                Some(Event::Task(task.unwrap())),
+                task.ok().map(Event::Task),
 
             recv(self.loader.receiver) -> task =>
-                Some(Event::Vfs(task.unwrap())),
+                task.ok().map(Event::Vfs),
 
             recv(self.flycheck_receiver) -> task =>
-                Some(Event::Flycheck(task.unwrap())),
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop.rs:181]: Unwraps are intentional here — suboinate threads should not die before we ask them to

**Comment by 0x5c** [comment]
I'm not familiar with the r-a release cycle, how long might this take to land in the VSCode extension?

**Comment by Veykril** [comment]
next monday for the stable release, tomorrow (~4hrs) for the nightly one

**Comment by lnicola** [comment]
@0x5c five days, or you can install a pre-release version tomorrow

**Pull request merged**